### PR TITLE
AEA-1113 build command for libp2p connection

### LINKED ISSUE #1710: `aea build` step in cli
**Is your feature request related to a problem? Please describe.**
Several packages contain go code which requires a build step.

**Describe the solution you'd like**
Allow package configs to specify a build step requirement and entry point. On call of `aea build` the packages with these requirements are identified and the build performed.

This prevents a situation where building takes place during launch of the agent and we can also avoid rebuilding unnecessarily.

### PR DESCRIPTION
**Pull request by solarw**
## Proposed changes

build command for libp2p connection

## Fixes

 #1710 #1618

## Types of changes

- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)

## Checklist

- [x] I have read the [CONTRIBUTING](../CONTRIBUTING.md) doc
- [x] I am making a pull request against the `develop` branch (left side). Also you should start your branch off our `develop`.
- [x] Lint and unit tests pass locally with my changes
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have checked that code coverage does not decrease.
- [x] I have checked that the documentation about the `aea cli` tool works
- [x] I have added necessary documentation (if appropriate)
- [x] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

File: aea/context/base.py [hunk 1]
```diff
@@ -85,6 +86,12 @@ def __init__(
         self._default_routing = default_routing
         self._storage_callable = storage_callable
         self._namespace = SimpleNamespace(**kwargs)
+        self._build_dir = build_dir
+
+    @property
+    def build_dir(self) -> str:
+        """Return storage instance if enabled in AEA."""
+        return self._build_dir
```
Comment (DavidMinarsch) [inline aea/context/base.py:94]: Why does this need to be exposed on context?

File: packages/fetchai/connections/p2p_libp2p/check_dependencies.py [hunk 2]
```diff
@@ -19,12 +19,35 @@
...
 from itertools import islice
 from typing import Any, Iterable, List, Pattern, Tuple
 
+from aea.helpers.base import ensure_dir
+
+
+try:
+    # flake8: noqa
+    # pylint: disable=unused-import,ungrouped-imports
+    from .connection import (  # type: ignore
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/check_dependencies.py:37]: Use full path: `from packages.fetchai.connections....`

File: packages/fetchai/connections/p2p_libp2p/connection.py [hunk 3]
```diff
@@ -294,25 +287,23 @@ async def start(self) -> None:
...
-                "Error while downloading golang dependencies and building it: {}\n{}".format(
-                    returncode, node_log
+                "Error while downloading golang dependencies and building it: {}".format(
+                    err_string
                 )
             )
         self.logger.info("Finished downloading golang dependencies.")
-
+        """
         # setup fifos
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/connection.py:307]: Remove unused code!

File: packages/fetchai/connections/p2p_libp2p/connection.py [hunk 4]
```diff
@@ -596,14 +586,26 @@ def __init__(self, **kwargs):
...
+        enforce(
+            self.configuration.build_directory,
+            "Connection Configuration build directory is not set!",
+        )
+        libp2p_node_module_path = os.path.join(
+            self.configuration.build_directory, LIBP2P_NODE_MODULE_NAME
+        )
+        enforce(
+            os.path.exists(libp2p_node_module_path),
+            f"Module {LIBP2P_NODE_MODULE_NAME} does not present in {self.configuration.build_directory}, please call `aea build` command",
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/connection.py:598]:
```suggestion
            f"Module {LIBP2P_NODE_MODULE_NAME} is not present in {self.configuration.build_directory}, please call `aea build` command and then try again.",
```

File: aea/aea.py [hunk 5]
```diff
@@ -184,6 +186,11 @@ def __init__(
 
         self._setup_loggers()
 
+    @classmethod
+    def get_build_dir(cls) -> str:
+        """Get agent build directory."""
+        return cls.DEFAULT_BUILD_DIR_NAME
```
Comment (DavidMinarsch) [inline aea/aea.py:192]: I don't see the point in defining this here.

**Review by DavidMinarsch — COMMENTED** [review]
Thanks, I have some concerns. Please let's remove any logic not strictly required (e.g. in context)

File: aea/context/base.py (hunk 1 above)
Comment (solarw) [reply aea/context/base.py:94]: thts correct, cane be reached by get_build_dir method

File: aea/aea.py (hunk 5 above)
Comment (solarw) [reply aea/aea.py:192]: will discuss later

File: packages/fetchai/connections/p2p_libp2p/connection.py (hunk 3 above)
Comment (solarw) [reply packages/fetchai/connections/p2p_libp2p/connection.py:307]: done

File: packages/fetchai/connections/p2p_libp2p/check_dependencies.py (hunk 2 above)
Comment (solarw) [reply packages/fetchai/connections/p2p_libp2p/check_dependencies.py:37]: will stay it as is

File: aea/configurations/base.py [hunk 6]
```diff
@@ -1124,6 +1124,17 @@ def __init__(
             build_entrypoint,
         )
         self.pypi_dependencies: Dependencies = dependencies if dependencies is not None else {}
+        self._build_directory = build_directory
+
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/configurations/base.py:1128]: done for every component configuration

File: aea/context/base.py [hunk 7]
```diff
@@ -49,6 +49,7 @@ def __init__(
         default_routing: Dict[PublicId, PublicId],
         search_service_address: Address,
         decision_maker_address: Address,
+        build_dir: str = ".build",
```
[the remark this answers is not in the document]
Comment (solarw) [reply aea/context/base.py:52]: done

File: tests/data/dummy_skill/skill.yaml [hunk 8]
```diff
@@ -14,7 +14,6 @@ fingerprint:
   handlers.py: QmYchfGvDXD5CxDHt6kChNsNbRxTcqViQTn65R7p8VJ4fu
   tasks.py: Qmegg4QsYSqSZN3q2zuRiBAToQ2LEiWrAPtUo7rCMrxjGJ
 fingerprint_ignore_patterns: []
-build_entrypoint: path/to/script.py
```
Comment (DavidMinarsch) [inline tests/data/dummy_skill/skill.yaml:17]: why removed?

File: tests/data/dummy_aea/aea-config.yaml [hunk 9]
```diff
@@ -6,7 +6,6 @@ license: Apache-2.0
 aea_version: '>=0.7.0, <0.8.0'
 fingerprint: {}
 fingerprint_ignore_patterns: []
-build_entrypoint: path/to/script.py
```
Comment (DavidMinarsch) [inline tests/data/dummy_aea/aea-config.yaml:9]: why removed?

File: packages/fetchai/connections/p2p_libp2p/connection.py [hunk 10]
```diff
@@ -299,24 +292,9 @@ async def start(self) -> None:
...
-            node_log = f.read()
-        if returncode != 0:
-            raise Exception(
-                "Error while downloading golang dependencies and building it: {}\n{}".format(
-                    returncode, node_log
-                )
-            )
-        self.logger.info("Finished downloading golang dependencies.")
+        self._log_file_desc.write("test")
+        self._log_file_desc.flush()
```
Comment (DavidMinarsch) [inline packages/fetchai/connections/p2p_libp2p/connection.py:269]: ?

File: tests/data/dummy_aea/aea-config.yaml (hunk 9 above)
Comment (DavidMinarsch) [inline tests/data/dummy_aea/aea-config.yaml:9]: Why removed?

**Review by DavidMinarsch — APPROVED** [review]
LGTM

File: tests/data/dummy_skill/skill.yaml (hunk 8 above)
Comment (DavidMinarsch) [reply tests/data/dummy_skill/skill.yaml:17]: Fails if actually tries to build

**Pull request merged**